File: 18809-Gaaaaaaaaaarden.cpp
#include <algorithm>
#include <cstring>
#include <functional>
#include <iostream>
#include <optional>
#include <queue>
#include <utility>
#include <vector>

typedef std::pair<int, int> point;

enum class state {
    LAKE,
    UNCULTIVABLE,
    CULTIVABLE,
    CULTIVATED,
    FLOURISHED
};

enum class color { GREEN, RED };

struct cell {
    state state = state::LAKE;
    std::optional<color> color = std::nullopt;
    std::optional<int> distance = std::nullopt;
};

int solve(cell[50][50], int, int, const std::vector<point>&, int, int);
int maximum_flowers(const cell[50][50], int, int, const std::vector<point>&);

int main()
{
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(0);

    int rows, columns, greens, reds;
    std::cin >> rows >> columns >> greens >> reds;

    cell garden[50][50];
    std::vector<point> cultivable_cells;

    for (int y = 0; y < rows; ++y) {
        for (int x = 0; x < columns; ++x) {
            int raw_state;
            std::cin >> raw_state;
            auto st = static_cast<state>(raw_state);
            garden[y][x].state = st;
            if (st == state::CULTIVABLE) {
                cultivable_cells.push_back(point(x, y));
            }
        }
    }

    std::cout
        << solve(garden, rows, columns, cultivable_cells, greens, reds)
        << '\n';

    return 0;
}

int solve(
    cell garden[50][50],
    int rows,
    int columns,
    const std::vector<point>& cultivable_cells,
    int greens,
    int reds
) {
    int result = -1;

    std::function<void(int,color,std::vector<point>::size_type)> inner;
    inner = [&](int left, auto color, auto progress) {
        if (left == 0) {
            switch (color) {
            case color::GREEN:
                inner(reds, color::RED, 0);
                break;
            case color::RED:
                result = std::max(
                    result,
                    maximum_flowers(garden, rows, columns, cultivable_cells)
                );
                break;
            }
        } else {
            for (auto i = progress; i < cultivable_cells.size(); ++i) {
                const auto cultivable = cultivable_cells[i];
                const int x = cultivable.first, y = cultivable.second;
                auto& cell = garden[y][x];
                if (cell.state != state::CULTIVABLE) { continue; }

                cell.state = state::CULTIVATED;
                cell.color = color;
                cell.distance = 0;

                inner(left - 1, color, i + 1);

                cell.state = state::CULTIVABLE;
                cell.color = std::nullopt;
                cell.distance = std::nullopt;
            }
        }
    };

    inner(greens, color::GREEN, 0);
    return result;
}

int maximum_flowers(
    const cell garden[50][50],
    int rows,
    int columns,
    const std::vector<point>& cultivable_cells
) {
    cell copy[50][50];
    std::memcpy(copy, garden, sizeof copy);

    auto visit = [&](point point, color color, int distance) -> bool {
        const int x = point.first, y = point.second;
        if (!(0 <= x && x < columns && 0 <= y && y < rows)) return false;

        auto& cell = copy[y][x];
        switch (cell.state) {
        case state::CULTIVABLE: case state::UNCULTIVABLE:
            cell.state = state::CULTIVATED;
            cell.color = color;
            cell.distance = distance;
            return true;

        case state::CULTIVATED: {
            const auto other_color = cell.color.value();
            const int other_distance = cell.distance.value();
            if (color != other_color && distance == other_distance) {
                cell.state = state::FLOURISHED;
            }
            return false;
        }

        case state::LAKE: case state::FLOURISHED:
            return false;
        }
    };

    auto adjs = [&](point p, color color, int distance) -> std::vector<point> {
        constexpr int dx[] = {1, 0, -1, 0}, dy[] = {0, 1, 0, -1};
        std::vector<point> result;
        for (int i = 0; i < 4; ++i) {
            const auto to_visit = point(p.first + dx[i], p.second + dy[i]);
            if (visit(to_visit, color, distance)) {
                result.push_back(to_visit);
            }
        }
        return result;
    };

    auto bfs = [&]() {
        std::queue<point> green_queue, red_queue;

        for (auto cell : cultivable_cells) {
            const auto curr = copy[cell.second][cell.first];
            if (curr.state == state::CULTIVATED) {
                switch (curr.color.value()) {
                case color::GREEN:
                    green_queue.push(cell);
                    break;
                case color::RED:
                    red_queue.push(cell);
                    break;
                }
            }
        }

        int current_distance = 0;
        while (!(green_queue.empty() && red_queue.empty())) {
            point front;

            auto predicate1 = [&]() -> bool {
                if (green_queue.empty()) { return false; }
                front = green_queue.front();
                const auto cell = copy[front.second][front.first];
                return cell.state == state::CULTIVATED &&
                    cell.distance.value() == current_distance;
            };
            auto predicate2 = [&]() -> bool {
                if (green_queue.empty()) { return false; }
                front = green_queue.front();
                const auto cell = copy[front.second][front.first];
                return cell.state == state::FLOURISHED;
            };
            auto predicate3 = [&]() -> bool {
                if (red_queue.empty()) { return false; }
                front = red_queue.front();
                const auto cell = copy[front.second][front.first];
                return cell.state == state::CULTIVATED &&
                    cell.distance.value() == current_distance;
            };

            if (predicate1()) {
                green_queue.pop();
                int new_distance = current_distance + 1;
                for (auto cell : adjs(front, color::GREEN, new_distance)) {
                    green_queue.push(cell);
                }
            } else if (predicate2()) {
                green_queue.pop();
            } else if (predicate3()) {
                red_queue.pop();
                int new_distance = current_distance + 1;
                for (auto cell : adjs(front, color::RED, new_distance)) {
                    red_queue.push(cell);
                }
            } else {
                ++current_distance;
            }
        }
    };

    bfs();

    int count = 0;
    for (int y = 0; y < rows; ++y) {
        for (int x = 0; x < columns; ++x) {
            if (copy[y][x].state == state::FLOURISHED) {
                ++count;
            }
        }
    }
    return count;
}

File: 18809-Gaaaaaaaaaarden.swift
// Swift is too slow for this problem
// Solved in C++ using exactly the same algorithm

struct Queue<Element> {
    private var array = [Element?](repeating: nil, count: 2500)

    private var headIndex = 0
    private var tailIndex = 0

    var isEmpty: Bool { headIndex == tailIndex }

    mutating func enqueue(_ newElement: Element) {
        array[tailIndex] = newElement
        if tailIndex == array.count - 1 {
            tailIndex = 0
        } else {
            tailIndex += 1
        }
    }

    mutating func dequeue() -> Element {
        let dequeued = array[headIndex]!
        if headIndex == array.count - 1 {
            headIndex = 0
        } else {
            headIndex += 1
        }
        return dequeued
    }

    var head: Element? { array[headIndex] }
}

enum Color { case green, red }

enum State: Equatable {
    case lake
    case cultivable
    case uncultivable
    case cultivated(Color, Int)
    case flourished

    init(int: Int) {
        switch int {
        case 0: self = .lake
        case 1: self = .uncultivable
        case 2: self = .cultivable
        default: fatalError()
        }
    }
}

struct Point: Equatable {
    let x: Int
    let y: Int

    var tuple: (Int, Int) { (x, y) }

    init(tuple: (Int, Int)) { (x, y) = tuple }

    init(x: Int, y: Int) { (self.x, self.y) = (x, y) }

    static let zero = Point(x: 0, y: 0)
}

func maximumFlowers(garden: [[State]], cultivableCells: [Point]) -> Int {
    var garden = garden
    let row = garden.count
    let column = garden[0].count

    func visit(_ point: Point, color: Color, distance: Int) -> Bool {
        guard (0..<row).contains(point.x) && (0..<column).contains(point.y)
            else { return false }

        let (x, y) = point.tuple
        switch garden[x][y] {
        case .cultivable, .uncultivable:
            garden[x][y] = .cultivated(color, distance)
            return true

        case let .cultivated(otherColor, otherDistance):
            if color != otherColor && distance == otherDistance {
                garden[x][y] = .flourished
            }
            return false

        case .lake, .flourished:
            return false
        }
    }

    func visitAdjacents(
        from point: Point,
        color: Color,
        distance: Int
    ) -> [Point] {
        [(1, 0), (0, 1), (-1, 0), (0, -1)]
            .map { Point(x: point.x + $0.0, y: point.y + $0.1) }
            .filter { visit($0, color: color, distance: distance) }
    }

    func bfs() {
        var greenQueue = Queue<Point>()
        var redQueue = Queue<Point>()
        for cell in cultivableCells {
            if case let .cultivated(color, _) = garden[cell.x][cell.y] {
                switch color {
                case .green: greenQueue.enqueue(cell)
                case .red: redQueue.enqueue(cell)
                }
            }
        }

        var currentDistance = 0
        while !greenQueue.isEmpty || !redQueue.isEmpty {
            if let head = greenQueue.head,
                case let .cultivated(_, distance) = garden[head.x][head.y],
                currentDistance == distance
            {
                let head = greenQueue.dequeue()
                visitAdjacents(
                    from: head,
                    color: .green,
                    distance: distance + 1
                )
                    .forEach { greenQueue.enqueue($0) }
            } else if let head = greenQueue.head,
                garden[head.x][head.y] == .flourished
            {
                let _ = greenQueue.dequeue()
            } else if let head = redQueue.head,
                case let .cultivated(_, distance) = garden[head.x][head.y],
                currentDistance == distance
            {
                let head = redQueue.dequeue()
                visitAdjacents(
                    from: head,
                    color: .red,
                    distance: distance + 1
                )
                    .forEach { redQueue.enqueue($0) }
            } else {
                currentDistance += 1
            }
        }
    }

    bfs()

    return garden.reduce(0) { $0 + $1.filter { $0 == .flourished }.count }
}

func solve(
    _ garden: inout [[State]],
    cultivableCells: [Point],
    greens: Int,
    reds: Int
) -> Int {
    var result = -1
    func inner(left: Int, color: Color, progress: Int) {
        if left == 0 {
            switch color {
            case .green:
                inner(left: reds, color: .red, progress: .zero)
            case .red:
                result = max(
                    result,
                    maximumFlowers(
                        garden: garden,
                        cultivableCells: cultivableCells
                    )
                )
            }
        } else {
            for i in progress..<cultivableCells.count {
                let cell = cultivableCells[i]
                guard garden[cell.x][cell.y] == .cultivable else { continue }
                garden[cell.x][cell.y] = .cultivated(color, 0)
                inner(left: left - 1, color: color, progress: i + 1)
                garden[cell.x][cell.y] = .cultivable
            }
        }
    }
    inner(left: greens, color: .green, progress: .zero)
    return result
}

let input = readLine()!.split(separator: " ").map { Int($0)! }
let row = input[0]
let column = input[1]
let greens = input[2]
let reds = input[3]

var garden = [[State]](
    repeating: [State](repeating: .lake, count: column),
    count: row
)
var cultivableCells = [Point]()

for i in 0..<row {
    readLine()!
        .split(separator: " ")
        .map({ State(int: Int($0)!) })
        .enumerated()
        .forEach {
            let (j, state) = $0
            garden[i][j] = state
            if state == .cultivable {
                cultivableCells.append(Point(x: i, y: j))
            }
        }
}

print(
    solve(
        &garden,
        cultivableCells: cultivableCells,
        greens: greens,
        reds: reds
    )
)
